### PR TITLE
[autoscaler] Service and Ingress per worker pod

### LINKED ISSUE #9042: [autoscaler] Service and Ingress resource per worker
Dear all

I'm trying to create an application which almost entirely runs on Ray. Currently it's running fine on a Ray cluster that I've deployed manually, but I would like to use the ray autoscaler (to pay less for GPU etc ;-)).
The application basically spawns ML pipelines in Ray actors and serves HTTP API servers, RTSP streams, etc.
In order to expose the communication endpoints to the outside world, the ray worker pods are exposed via kubernetes services and these services are exposed via Ingress resources.
My current deployment looks like this:
* A ray head and a few worker pods
* A kubernetes service _for each worker_ exposing port range 7000->7010 on each respective worker
* Ingress rules for each service, exposing the ports like: 
<details>
  <summary>ingress translation</summary>

| ingress rule | service:port |
| ------------- | ------------- |
| https://domainname.com/worker1/7000 | ray-worker-1:7000 | 
| https://domainname.com/worker1/... | ray-worker-1:... | 
| https://domainname.com/worker1/7010 | ray-worker-1:7010 | 
| https://domainname.com/worker2/7000 | ray-worker-2:7000 | 
| https://domainname.com/worker2/... | ray-worker-2:... | 
| https://domainname.com/worker2/7010 | ray-worker-2:7010 | 
| .... | .... |

</details>
When e.g. an HTTP API server is spawned inside an actor, it knows the port it's running on (e.g. 7000) and knows the worker it's running in (currently via the HOSTNAME env var), so based on this information it can determine on which URL the API server is exposed for end users.

I'd like to use the same architecture with autoscale, but I'm missing the possibility to deploy/remove a kubernetes service and ingress resource along with each pod. To accomplish this, the autoscale yaml could be like this:
<details>
  <summary>autoscale.yaml</summary>

```
provider:
    type: kubernetes
# Kubernetes pod config for the head node pod.
head_node:
    pod:
        apiVersion: v1
        kind: Pod
        metadata:
            # Automatically generates a name for the pod with this prefix.
            generateName: ray-head-
            labels:
                component: ray-head
        spec:
    service:  # optional
        # Service that maps to the head node of the Ray cluster.
        apiVersion: v1
        kind: Service
        metadata:
    ingress:  # optional
        apiVersion: extensions/v1beta1
        kind: Ingress
        metadata:
# Kubernetes pod config for worker node pods.
worker_nodes:
    pod:
        apiVersion: v1
        kind: Pod
        metadata:
            # Automatically generates a name for the pod with this prefix.
            generateName: ray-worker-
        labels:
            component: 
    service:  # optional
        # Service that maps to the head node of the Ray cluster.
        apiVersion: v1
        kind: Service
        metadata:
    ingress:  # optional
        apiVersion: extensions/v1beta1
        kind: Ingress
        metadata:
```

</details>

The only thing that's not 100% straightforward is that there should be some intelligence in the autoscaler to map the right service on the right worker. E.g. by assigning a unique label to every pod and using this label as selector for the service, so there is a 1:1 relationship between service and pod.

I hope my goal is clear and I think it'll be useful functionality for other people too.
Thanks for taking this into discussion/consideration!

BR, Pieterjan

CC @edoakes

### PR DESCRIPTION
**Pull request by PidgeyBE**
## Why are these changes needed?

Currently it's not possible to add services and ingress rules per ray worker pod on kubernetes.
The changes in this PR allow to not only define a pod manifest, but also a service and ingress manifest for workers.
Each time a worker pod is scaled up/down, a matching service and ingress resource is also scaled up/down.

## Related issue number

Closes https://github.com/ray-project/ray/issues/9042

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by rkooo567** [comment]
cc @edoakes

**Comment by edoakes** [comment]
Thanks @PidgeyBE this looks pretty reasonable to me. I'm wondering if it might make more sense to have a "magic" variable that gets autofilled by the autoscaler instead of just specifically filling in the `serviceName` field. Instead of not specifying the `serviceName`, users would specify it is `serviceName: ${RAY_POD_NAME}` or something like that. Thoughts?

File: python/ray/autoscaler/kubernetes/node_provider.py [hunk 1]
```diff
@@ -97,3 +144,25 @@ def get_command_runner(self,
...
+    """
+    if isinstance(spec, dict):
+        dict_keys = list(spec.keys())
+        for k in dict_keys:
+            spec[k] = _add_service_name_to_service_port(spec[k], svc_name)
+
+            # Every place that contains a servicePort definition should also
+            # have the serviceName defined.
+            if k == "servicePort":
+                spec["serviceName"] = svc_name
```
Comment (edoakes) [inline python/ray/autoscaler/kubernetes/node_provider.py:161]: We should probably raise an exception here (or at least print a warning) if the user specified the `serviceName`

**Comment by PidgeyBE** [comment]
> Thanks @PidgeyBE this looks pretty reasonable to me. I'm wondering if it might make more sense to have a "magic" variable that gets autofilled by the autoscaler instead of just specifically filling in the `serviceName` field. Instead of not specifying the `serviceName`, users would specify it is `serviceName: ${RAY_POD_NAME}` or something like that. Thoughts?
I agree that's a better way of doing it! I'll update the code!

File: python/ray/autoscaler/kubernetes/node_provider.py (hunk 1 above)
Comment (PidgeyBE) [reply python/ray/autoscaler/kubernetes/node_provider.py:161]: Good idea! I'll do that!

**Comment by PidgeyBE** [comment]
@edoakes I've addressed you comments!

File: python/ray/autoscaler/kubernetes/node_provider.py [hunk 2]
```diff
@@ -97,3 +144,31 @@ def get_command_runner(self,
...
+            spec[k] = _add_service_name_to_service_port(spec[k], svc_name)
+
+            # The magic string ${RAY_POD_NAME} is replaced with
+            # the true service name, which is equal to the worker pod name.
+            if k == "serviceName":
+                if spec[k] != "${RAY_POD_NAME}":
+                    raise ValueError(
+                        "The value of serviceName must be set to "
+                        "${RAY_POD_NAME}. It is automatically replaced "
+                        "when autoscaling.")
```
Comment (edoakes) [inline python/ray/autoscaler/kubernetes/node_provider.py:165]:
```suggestion
                        "when using the autoscaler.")
```

File: python/ray/autoscaler/kubernetes/example-ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            kubernetes.io/ingress.class: nginx
+        spec:
+          rules:
+            - host: localhost
+              http:
+                paths:
+                - path: /
+                  backend:
+                    # The value of the serviceName will be
+                    # automatically set to the name of the pod
```
Comment (edoakes) [inline python/ray/autoscaler/kubernetes/example-ingress.yaml:293]:
```suggestion
                    # The value of the serviceName must be set to `${RAY_POD_NAME} and will be
                    # automatically replaced by the name of the pod.
```

**Review by edoakes — APPROVED** [review]
Ok, looks reasonable to me. In the future we may want to also support replacing `RAY_POD_NAME` in other places in the spec but for now this seems like the only place it's necessary.

**Pull request merged**